core/src/main/java/org/apache/iceberg/Snapshots.java

```diff
@@ -0,0 +1,226 @@
+      List<ManifestFile> manifests = isStarting ? snapshot.dataManifests() :
+          snapshot.dataManifests().stream().filter(m -> m.snapshotId().equals(snapshot.snapshotId()))
+              .collect(Collectors.toList());
+
+      List<Pair<ManifestFile, Integer>> manifestIndexes = indexManifests(manifests);
+      List<Pair<ManifestFile, Integer>> skippedManifestIndexes = skipManifests(manifestIndexes, startFileIndex);
+
+      return generateMicroBatch(skippedManifestIndexes, startFileIndex, targetSizeInBytes, isStarting);
+    }
+
+    private List<Pair<ManifestFile, Integer>> indexManifests(List<ManifestFile> manifestFiles) {
+      int currentFileIndex = 0;
+      List<Pair<ManifestFile, Integer>> manifestIndexes = Lists.newArrayList();
+
+      for (ManifestFile manifest : manifestFiles) {
+        int filesCount = manifest.addedFilesCount() + manifest.existingFilesCount();
+        manifestIndexes.add(Pair.of(manifest, currentFileIndex));
+        currentFileIndex += filesCount;
+      }
+
+      return manifestIndexes;
+    }
+
+    private List<Pair<ManifestFile, Integer>> skipManifests(List<Pair<ManifestFile, Integer>> indexedManifests,
+                                                            int startFileIndex) {
+      if (startFileIndex == 0) {
+        return indexedManifests;
+      }
+
+      int index = 0;
+      for (Pair<ManifestFile, Integer> manifest : indexedManifests) {
+        if (manifest.second() > startFileIndex) {
+          break;
+        }
+
+        index++;
+      }
+
+      return indexedManifests.subList(index - 1, indexedManifests.size());
+    }
+
+    private MicroBatch generateMicroBatch(List<Pair<ManifestFile, Integer>> indexedManifests,
+                                          int startFileIndex, long targetSizeInBytes, boolean isStarting) {
+      if (indexedManifests.isEmpty()) {
+        return new MicroBatch(snapshot.snapshotId(), startFileIndex, startFileIndex + 1, 0L,
+            CloseableIterable.empty(), true);
+      }
+
+      long currentSizeInBytes = 0L;
+      int currentFileIndex = 0;
+      List<CloseableIterable<FileScanTask>> batchTasks = Lists.newArrayList();
+
+      for (Pair<ManifestFile, Integer> pair : indexedManifests) {
+        currentFileIndex = pair.second();
+
+        CloseableIterable<FileScanTask> tasks;
+        if (isStarting) {
+          tasks = full(ImmutableList.of(pair.first()));
+        } else {
+          tasks = appends(ImmutableList.of(pair.first()));
+        }
+
+        List<FileScanTask> batch = Lists.newArrayList();
+
+        for (FileScanTask task : tasks) {
+          if (currentFileIndex < startFileIndex) {
+            currentFileIndex++;
+            continue;
+          }
+
+          if (currentSizeInBytes + task.length() <= targetSizeInBytes) {
+            batch.add(task);
+            currentSizeInBytes += task.length();
+          } else {
+            batchTasks.add(CloseableIterable.combine(batch, tasks));
```

Comment:
I don't think the use of `CloseableIterable` is correct in this class. There is no need to return a closeable collection because the Iterable, tasks, has already been used. Rather than opening a reader, using its contents, and then returning a collection that will close the reader, just close the reader here instead.

Another problem is that creating multiple `CloseableIterables` with the same underlying resource to close doesn't make sense. The first group will close the reader and the others will re-close the same one without actually using it.

Instead of using `CloseableIterable`, this should just use a `List` for the tasks in a micro batch.

Reply:
Changed.